library/std/src/sys/xous/thread.rs

```diff
@@ -0,0 +1,141 @@
+
+        // Allocate the whole thing, then divide it up after the fact. This ensures that
+        // even if there's a context switch during this function, the whole stack plus
+        // guard pages will remain contiguous.
+        let stack_plus_guard_pages: &mut [u8] = map_memory(
+            None,
+            None,
+            stack_size + GUARD_PAGE_SIZE + GUARD_PAGE_SIZE,
+            MemoryFlags::R | MemoryFlags::W | MemoryFlags::X,
+        )
+        .map_err(|code| io::Error::from_raw_os_error(code as i32))?;
+
+        // No access to this page. Note: Write-only pages are illegal, and will
+        // cause an access violation.
+        update_memory_flags(&mut stack_plus_guard_pages[0..GUARD_PAGE_SIZE], MemoryFlags::W)
+            .map_err(|code| io::Error::from_raw_os_error(code as i32))?;
+
+        // No access to this page. Note: Write-only pages are illegal, and will
+        // cause an access violation.
+        crate::os::xous::ffi::update_memory_flags(
+            &mut stack_plus_guard_pages[(GUARD_PAGE_SIZE + stack_size)..],
+            MemoryFlags::W,
+        )
+        .map_err(|code| io::Error::from_raw_os_error(code as i32))?;
+
+        let tid = create_thread(
+            thread_start as *mut usize,
+            &stack_plus_guard_pages[GUARD_PAGE_SIZE..(stack_size - GUARD_PAGE_SIZE)],
+            p as usize,
+            stack_plus_guard_pages.as_ptr() as usize,
+            stack_size,
+            0,
+        )
+        .map_err(|code| io::Error::from_raw_os_error(code as i32))?;
+
+        extern "C" fn thread_start(main: *mut usize, guard_page_pre: usize, stack_size: usize) {
+            unsafe {
+                // Finally, let's run some code.
+                Box::from_raw(main as *mut Box<dyn FnOnce()>)();
+            }
+
+            // Destroy TLS, which will free the TLS page and call the destructor for
+            // any thread local storage.
+            unsafe {
+                crate::sys::thread_local_key::destroy_tls();
+            }
+
+            // Deallocate the stack memory, along with the guard pages.
+            let mapped_memory_base = guard_page_pre;
+            let mapped_memory_length = GUARD_PAGE_SIZE + stack_size + GUARD_PAGE_SIZE;
+            unsafe {
+                asm!(
+                    "ecall",
+                                        inlateout("a0") Syscall::UnmapMemory as usize => _,
+                                        inlateout("a1") mapped_memory_base => _,
+                                        inlateout("a2") mapped_memory_length => _,
+                                        options(nomem, nostack)
+                );
+            }
+
+            // Exit the thread by returning to the magic address 0xff80_3000usize,
+            // which tells the kernel to deallcate this thread.
+            unsafe {
+                asm!("ret", in("a0") 0, in("ra") 0xff80_3000usize,
+                    options(nomem, nostack, noreturn)
+                );
```

Comment:
This asm block will need to be merged with the previous one. asm blocks can be implemented as function calls (eg my cranelift based backend does this because cranelift doesn't have native inline asm support), which doesn't work if the stack is deallocated.